lib/puppet/http/service.rb

```diff
@@ -43,7 +43,19 @@ def connect(ssl_context: nil)
 
   def add_puppet_headers(headers)
     modified_headers = headers.dup
+
+    # Add 'X-Puppet-Profiling' to enable performance profiling if turned on
     modified_headers['X-Puppet-Profiling'] = 'true' if Puppet[:profile]
+
+    # Add additional user-defined headers if they are defined
+    Puppet[:http_extra_headers].each do |name, value|
+      if modified_headers.key?(name)
```

Comment:
Since http header names are case-insensitive, you'll need to do something like the following. The number of extra headers should be fairly low, so I'm not too concerned about `O(n^2)`

```ruby
modifies_headers.keys.find { |key| key.casecmp(name) == 0 }
```

Reply:
I would love to add a test for `add_puppet_headers` to`spec/unit/http/service_spec.rb`, but I can't since it's a protected method. Does the benefit of having it protected outweigh being able to test the method directly? I'm mostly being lazy and not wanting to duplicate functionality tests everywhere. I'm adding a check here to raise if we have a header with an empty value

Reply:
In that case I'd define a test service subclass like `class TestService < Puppet::HTTP::Service` and define a method `headers` that calls the protected `add_puppet_headers`. Then in your test call `TestService#headers` and verify it parses/rejects headers as expected. That way you can everything in one place, instead of once per service.

Reply:
in `spec/unit/http/service_spec.rb` or a different file?

Reply:
yeah in `service_spec.rb`. We do something similar in `session_spec.rb` when creating the `DummyResolver`.